File: 01/solver.py
import sys


FILE = sys.argv[1] if len(sys.argv) > 1 else "input.txt"

with open(FILE) as f:
    PUZZLE_INPUT = f.read()

lines = [line.strip() for line in PUZZLE_INPUT.split("\n") if line]

result = 0

for l in lines:
    numbers = []
    for ch in l:
        if ch.isdigit():
            numbers.append(ch)
    first_last = numbers[0] + numbers[~0]
    result += int(first_last)

# Part 1 = 54644
print(f"answer = {result}")

result = 0

converter = {
    "one": "1",
    "two": "2",
    "three": "3",
    "four": "4",
    "five": "5",
    "six": "6",
    "seven": "7",
    "eight": "8",
    "nine": "9"
    }

for l in lines:
    temp = ""
    numbers = []
    for ch in l:
        if ch.isdigit():
            numbers.append(ch)
        else:
            temp += ch
            for n, v in converter.items():
                if temp.endswith(n):
                    numbers.append(v)
                    break

    first_last = numbers[0] + numbers[~0]
    result += int(first_last)

# Part 2 = 53348
print(f"answer = {result}")

File: 01/solver.exs
file = "input.txt"

input_data =
  File.stream!(file)
  |> Stream.map(&String.replace(&1, "\n", ""))
  |> Enum.to_list()

# |> IO.inspect()

result =
  input_data
  |> Enum.reduce(0, fn line, result ->
    {first, second} =
      line
      |> String.graphemes()
      |> Enum.reduce({}, fn ch, acc ->
        cond do
          String.match?(ch, ~r"\d") and tuple_size(acc) == 0 ->
            as_int = String.to_integer(ch)
            {as_int, as_int}

          String.match?(ch, ~r"\d") ->
            {elem(acc, 0), String.to_integer(ch)}

          true ->
            acc
        end
      end)

    result + first * 10 + second
  end)

IO.puts("Answer to part 1 = #{result}")

digit_strings = [
  "one",
  "two",
  "three",
  "four",
  "five",
  "six",
  "seven",
  "eight",
  "nine"
]

starts_with_number_string = fn str ->
  digit_strings
  |> Enum.with_index(1)
  |> Enum.reduce(nil, fn {number, index}, acc ->
    if String.starts_with?(str, number) do
      index
    else
      acc
    end
  end)
end

update_first_last = fn
  current, nil -> current
  {}, value -> {value, value}
  {f, _}, value -> {f, value}
end

result =
  input_data
  |> Enum.reduce(0, fn line, result ->
    {first, second} =
      line
      |> String.graphemes()
      |> Enum.with_index()
      |> Enum.reduce({}, fn {ch, i}, acc ->
        value =
          cond do
            String.match?(ch, ~r"\d") ->
              String.to_integer(ch)

            string_digit = starts_with_number_string.(String.slice(line, i..-1)) ->
              string_digit

            true ->
              nil
          end

        update_first_last.(acc, value)
      end)

    result + first * 10 + second
  end)

IO.puts("Answer to part 2 = #{result}")
